lib/Backends/OpenCL/OpenCLDeviceManager.cpp

```diff
@@ -434,24 +434,51 @@ void OpenCLDeviceManager::copyInputsToDevice(
     auto symbolInfo = it->second;
     auto addr = symbolInfo.offset;
     auto numBytes = PH.second->getUnpaddedSizeInBytes();
-    // Issue a non-blocking command to copy the buffer to the device.
-    auto buf = PH.second->getUnsafePtr();
     cl_event event{nullptr};
 
-    cl_int err = clEnqueueWriteBuffer(
-        devBindings->commandQueue, devBindings->deviceBuffer,
-        /* blocking_write */ CL_FALSE, addr, numBytes, buf,
-        /* num_events_in_wait_list */ 0,
-        /* event_list */ nullptr,
-        /* event */ profilingEnabled ? &event : nullptr);
-    CHECK_EQ(err, CL_SUCCESS) << "Unable to copy data to the device";
-    if (profilingEnabled) {
-      std::string name = ("(H2D) " + PH.first->getName()).str();
-      devBindings->kernelLaunches.emplace_back(name, "copy", event);
+    auto *tensor = PH.second;
+    // If the tensor is a host tensor --> assign a tensor buffer and copy to
+    // the device.
+    if (!tensor->isDeviceResident()) {
+      // Host tensor --> copy to device in the corresponding offset in
+      // contiguous memory.
+      OpenCLDeviceTransferContext *transferContext =
+          new OpenCLDeviceTransferContext(devBindings->deviceBuffer, addr);
+      transferToDevice(*tensor, transferContext);
+    } else { // Tensor is device resident
+      DeviceResidencyInfo info = tensor->getDeviceResidencyInfo();
+      // If the tensor resides on this device, copy on the device.
+      // else: transter tensor to this device via the host
+      if (info.deviceManager == this) {
+        OpenCLDeviceTransferContext *ctx =
+            static_cast<OpenCLDeviceTransferContext *>(info.context);
+        cl_int err = clEnqueueCopyBuffer(
+            devBindings->commandQueue, ctx->tensorBuffer,
+            devBindings->deviceBuffer, ctx->offset, addr, numBytes,
+            /* num_events_in_wait_list */ 0,
+            /* event_list */ nullptr,
+            /* event */ profilingEnabled ? &event : nullptr);
+        CHECK_EQ(err, CL_SUCCESS) << "Unable to copy data from the device";
+        if (profilingEnabled) {
+          std::string name = ("(D2D) " + PH.first->getName()).str();
+          devBindings->kernelLaunches.emplace_back(name, "copy", event);
+        }
+        clFinish(devBindings->commandQueue);
+        // release temporary cl_mem buffer
+        releaseDeviceTensor(*tensor);
+        tensor->moveToDevice(
+            (DeviceManager *)this,
+            new OpenCLDeviceTransferContext(devBindings->deviceBuffer, addr));
+        delete ctx;
```

Comment:
This is a bit confusing, if it's already on the right device why do we have to do anything?

Reply:
Oh is this to handle the contiguous memory ranges?

Reply:
Yes - this is the temporary buffering to workaround compiled function contiguous memory buffer.
Refactoring this "buffering" code (in both copyInputs, copyOutputs) to a helper function.